lib/core/include/irods/irods_load_plugin.hpp

```diff
@@ -94,6 +101,274 @@ namespace irods
+        using logger = irods::experimental::log::server;
+
+        int rtld_lazy_or_now = RTLD_LAZY;
+        int rtld_local_or_global = RTLD_LOCAL;
+        int rtld_nodelete = 0;
+
+        try {
+            std::ifstream lib(_plugin_path, std::ios::binary);
+            if (!lib) {
+                // clang-format off
+                logger::error({
+                    {"plugin_path", _plugin_path},
+                    {"message", "Could not parse DT_FLAGS_1: failed to open plugin file"}
+                });
+                // clang-format on
+                // NOLINTNEXTLINE(hicpp-signed-bitwise)
+                return rtld_lazy_or_now | rtld_local_or_global | rtld_nodelete;
+            }
+
+            // make sure we're actually reading an ELF file
+            const std::array<char, 4> elf_mag{{ELFMAG0, ELFMAG1, ELFMAG2, ELFMAG3}};
+            std::array<char, 4> elf_mag_in{};
+            // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+            lib.read(reinterpret_cast<char*>(elf_mag_in.data()), elf_mag_in.size());
+            if (elf_mag != elf_mag_in) {
+                // clang-format off
+                logger::error({
+                    {"plugin_path", _plugin_path},
+                    {"message", "Could not parse DT_FLAGS_1: invalid file format (ELF magic number mismatch)"}
+                });
+                // clang-format on
+                // NOLINTNEXTLINE(hicpp-signed-bitwise)
+                return rtld_lazy_or_now | rtld_local_or_global | rtld_nodelete;
+            }
+
+            // get ELFCLASS
+            std::uint8_t ei_class{};
+            // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+            lib.read(reinterpret_cast<char*>(&ei_class), sizeof(ei_class));
+
+            lib.seekg(0);
+
+            std::uint64_t e_shoff{};
+            std::uint16_t e_shentsize{};
+            std::uint64_t e_shnum{};
+            std::uint32_t e_shstrndx{};
+            if (ei_class == ELFCLASS32) {
+                Elf32_Ehdr elf_hdr_in{};
+                // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+                lib.read(reinterpret_cast<char*>(&elf_hdr_in), sizeof(elf_hdr_in));
+                e_shoff = elf_hdr_in.e_shoff;
+                e_shentsize = elf_hdr_in.e_shentsize;
+                e_shnum = elf_hdr_in.e_shnum;
+                e_shstrndx = elf_hdr_in.e_shstrndx;
+            }
+            else if (ei_class == ELFCLASS64) {
+                Elf64_Ehdr elf_hdr_in{};
+                // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+                lib.read(reinterpret_cast<char*>(&elf_hdr_in), sizeof(elf_hdr_in));
+                e_shoff = elf_hdr_in.e_shoff;
+                e_shentsize = elf_hdr_in.e_shentsize;
+                e_shnum = elf_hdr_in.e_shnum;
+                e_shstrndx = elf_hdr_in.e_shstrndx;
+            }
+            else {
+                // clang-format off
+                logger::warn({
```

Comment:
confirming this is `warn` on purpose.

Reply:
I used `warn` because the situation is less severe than than the cases we use `error` for above, but either one makes sense here to me.

Reply:
got it - up to you.